plug support of CCDB in serve-d

rtbo opened the pull request:
WIP to enable CCDB in serve-d:
 - new config key: `d.ccdbPath` for the path to `compile_commands.json`
 - lazy load ccdb component if `d.ccdbPath` is set (in `onConfigChange` event)

Works already well like this.
I'd like to go further: running the command associated with the file to parse for diagnostics (instead of DScanner).

In ideal world, the import paths, versions etc. would not be a global list for the workspace, but a different set for each file.
For this I need one dcd-server instance per open file right?
For import paths it would not matter much, but for versions we can imagine at some point to grey out code related to unset versions.

WebFreak001 commented on workspace-d/source/workspaced/com/ccdb.d:
```diff
@@ -21,61 +19,51 @@ class ClangCompilationDatabaseComponent : ComponentWrapper
-		// we release memory when possible to avoid having at the same time more than
-		// two represention of the same data
-		jsonString = null;
+		trace("parsing CCDB");
```
```suggestion
		trace("parsing CCDB from ", filename);
```

WebFreak001 commented on workspace-d/source/workspaced/com/ccdb.d:
```diff
@@ -21,61 +19,51 @@ class ClangCompilationDatabaseComponent : ComponentWrapper
+		stringImportPathProvider = &stringImports;
+		importFilesProvider = &fileImports;
+		projectVersionsProvider = &versions;
+		debugSpecificationsProvider = &debugVersions;
```
why remove the ability to enable/disable the providers?

also this should call
```d
if (auto dbPath = config.get!string("ccdb", "dbPath", null))
	loadDb(dbPath);
```

to load dbPath immediately, if set.

rtbo replied:
I had errors due to missing workspace instance. I'm only starting to get my head around the code base. 
I don't quite get yet how this configuration works, and when/where all the instances are created.
I will correct this as soon as I understand how these things work together.

WebFreak001 commented on source/served/extension.d:
```diff
@@ -111,6 +112,11 @@ void changedConfig(ConfigWorkspace target, string[] paths, served.types.Configur
+		case "d.ccdbPath":
+			if (backend.has!ClangCompilationDatabaseComponent(workspaceFs))
+				backend.get!ClangCompilationDatabaseComponent(workspaceFs)
+					.setPath(config.d.ccdbPath.userPath);
```
this change event will only trigger if the config changes from the default - so it would probably be a good idea to also include the config value in the CCDB component config.

Additionally the config change event may trigger before the instance is loaded, so the setPath function would never be run there with the current implementation.

WebFreak001 commented on source/served/extension.d:
```diff
@@ -353,6 +359,8 @@ void doGlobalStartup()
 		trace("Registering fsworkspace");
 		backend.register!FSWorkspaceComponent(false);
+		trace("Registering ccdb");
+		backend.register!ClangCompilationDatabaseComponent;
```
this is being auto-registered, meaning it will override the DUB and FSWorkspace components - add the `false` argument to disable auto-registration and add code to enable it conditionally per-workspace (in function `delayedProjectActivation`) where it is actually used.

rtbo replied:
OK. I started with `false` but it didn't do anything.
I will look at `delayedProjectActivation`.

WebFreak001 commented on source/served/extension.d:
```diff
@@ -798,8 +848,8 @@ string determineOutputFolder()
-			return buildPath(environment["HOME"], ".local", "share", "code-d", "bin");
-		else
+		return buildPath(environment["HOME"], ".local", "share", "code-d", "bin");
+	else
```
broken formatting

WebFreak001 commented on source/served/extension.d:
```diff
@@ -426,7 +434,7 @@ struct RootSuggestion
 
 RootSuggestion[] rootsForProject(string root, bool recursive, string[] blocked,
-		string[] extra)
+string[] extra)
```
broken formatting

WebFreak001 commented on source/served/extension.d:
```diff
@@ -991,9 +1042,43 @@ void onDidSaveDocument(DidSaveTextDocumentParams params)
 }
 
+@protocolNotification("workspace/didChangeWatchedFiles")
+void onDidChangeWatchedFiles(DidChangeWatchedFilesParams params)
```
this function name is too generic, it should be called something with CCDB as it only handles that.

Additionally this function and `discoverCcdb` should be moved into a separate ccdb module to avoid cluttering extension with new features that could be nicely grouped together. When you create a separate module, you need to `static import` it in extension.d (see the top of the file) and add it to the list of members at the bottom of the file

WebFreak001 commented on source/served/extension.d:
```diff
@@ -991,9 +1042,43 @@ void onDidSaveDocument(DidSaveTextDocumentParams params)
+{
+	foreach (c; params.changes)
+	{
+		trace("watched file did change: ", c);
```
you should check if `c` is actually the filename you look for and not some other file like profilegc.log

WebFreak001 commented on source/served/extension.d:
```diff
@@ -633,42 +657,70 @@ void delayedProjectActivation(WorkspaceD.Instance instance, string workspaceRoot
+		didLoadDubProject();
+
+	string ccdbPath = proj.config.d.ccdbPath;
+	if (!ccdbPath && !loadedDub)
```
```suggestion
	if (!ccdbPath.length && !loadedDub)
```

(otherwise if ccdbPath is empty string, it evaluates to true)

WebFreak001 commented on source/served/extension.d:
```diff
@@ -633,42 +657,70 @@ void delayedProjectActivation(WorkspaceD.Instance instance, string workspaceRoot
+	string ccdbPath = proj.config.d.ccdbPath;
+	if (!ccdbPath && !loadedDub)
+		ccdbPath = discoverCcdb(workspaceRoot);
+	if (ccdbPath)
```
```suggestion
	if (ccdbPath.length)
```

WebFreak001 commented on workspace-d/source/workspaced/com/ccdb.d:
```diff
@@ -34,48 +35,58 @@ class ClangCompilationDatabaseComponent : ComponentWrapper
+	{
+		import std.path : buildNormalizedPath;
+
+		if (dbPath)
```
```suggestion
		if (dbPath.length)
```

WebFreak001 commented on source/served/extension.d:
```diff
@@ -633,42 +656,41 @@ void delayedProjectActivation(WorkspaceD.Instance instance, string workspaceRoot
+		error("Failed starting dub or CCDB in ", root, " - falling back to fsworkspace");
+		proj.startupError(workspaceRoot, translate!"d.ext.dubFail"(instance.cwd, err ? err.msg : ""));
+	}
+	try
```
you moved the `try` with fsworkspace out of the `if (!loadedDub)` check.

The error message was in a further `if (!disableDub)` to only show the error when DUB was enabled and actually attempted to be loaded. This should probably be changed to only load fsworkspace again if neither dub nor ccdb are loaded. (otherwise fsworkspace might override it)

This is probably the last thing I have to note for this PR, I still need to test out different projects of dub/fsworkspace/ccdb locally.

rtbo replied:
Is it better with 43e2bbdb4833f68eda6ce7730d6d563df9784493?
I'm not too sure what the fallback logic should be. Do I need to add a `d.ext.ccdbFail` message? (and load the db in a try/catch)

WebFreak001 replied:
yeah that commit is better I think. A proper `ccdbFail` message would be nice for the user

rtbo commented:
Hi, I thought to remember this was merged, but no !
Nevertheless I rebased on master and added linting diagnostics. Works pretty-well and is much faster than Dub's as only a single file is compiled each time
.
![image](https://github.com/Pure-D/serve-d/assets/3889105/9802f051-b444-4bd9-bd87-f3f543b823fe)

WebFreak001 commented:
yeah sorry about the delay, could you also auto-format your ccdb files? There seems to be mixed spaces and tabs occasionally.
